src/runtime/xrt.cpp

```diff
@@ -0,0 +1,628 @@
+        xrtKernelClose(state->handle);
+
+        free(state);
+    }
+}
+
+WEAK int halide_xrt_run(void *user_context,
+                        void *state_ptr,
+                        const char *entry_name,
+                        halide_type_t arg_types[],
+                        void *args[],
+                        int8_t arg_is_buffer[]) {
+    int ret;
+    XrtKernelState *state;
+    xrtRunHandle run_handle;
+    uint32_t num_args;
+    uint64_t t_before, t_after;
+
+    debug(user_context)
+        << "XRT: halide_xrt_run (user_context: " << user_context << ", "
+        << "entry: " << entry_name << ")\n";
+
+    XRTContext context(user_context);
+    if (context.error_code)
+        return context.error_code;
+
+    state = (XrtKernelState *)state_ptr;
+    run_handle = xrtRunOpen(state->handle);
+    if (run_handle == XRT_NULL_HANDLE) {
+        error(user_context)
+            << "XRT: halide_xrt_run: "
+            << "failed to open run handle for kernel: " << entry_name << "\n";
+        return halide_error_code_generic_error;
+    }
+
+    num_args = 0;
+    while (args[num_args] != nullptr) {
+        static const char *const type_code_names[] = {
+            "int",
+            "uint",
+            "float",
+            "handle",
+            "bfloat",
+        };
+
+        debug(user_context)
+            << "XRT: halide_xrt_run: "
+            << "arg[" << num_args << "]: "
+            << (arg_is_buffer[num_args] ? "buffer" : "scalar") << ", "
+            << "type: " << type_code_names[arg_types[num_args].code] << "\n";
+
+        if (arg_is_buffer[num_args]) {
+            halide_buffer_t *buffer = (halide_buffer_t *)args[num_args];
+            XrtBufferHandle *buf = (XrtBufferHandle *)buffer->device;
+
+            // Buffer not yet allocated. Allocate it now.
+            if (buf->handle == XRT_NULL_HANDLE) {
+                buf->handle = xrtBOAlloc(context.device, buf->size, XRT_BO_FLAGS_CACHEABLE,
+                                         (xrtMemoryGroup)xrtKernelArgGroupId(state->handle, num_args));
+                if (buf->handle == XRT_NULL_HANDLE) {
+                    error(user_context)
+                        << "XRT: halide_xrt_run: "
+                        << "failed to allocate buffer with size: " << (uint64_t)buf->size
+                        << " for kernel: " << entry_name << "\n";
+                    xrtRunClose(run_handle);
+                    return halide_error_code_generic_error;
+                }
+
+                debug(user_context) << "XRT: halide_xrt_run: "
+                                    << "allocated buffer with size: " << (uint64_t)buf->size
+                                    << " at physical address: " << (void *)xrtBOAddress(buf->handle) << "\n";
+
+                if (buf->copy_to_device_pending) {
+                    debug(user_context) << "  buffer has a copy to device pending.\n";
+                    // memset(buffer->host, 0xFF, buf->size);
```

Comment:
Don't leave in commented-out code -- either delete it, or add a comment about why it's intentionally commented out (eg for debugging)